File: java/src/main/java/com/sketches_oxide/membership/CuckooFilter.java
package com.sketches_oxide.membership;

import com.sketches_oxide.NativeSketch;
import com.sketches_oxide.MergeableSketch;
import com.sketches_oxide.native.SketchOxideNative;

/**
 * CuckooFilter - Space-Efficient Probabilistic Membership with Deletion.
 *
 * A modern alternative to Bloom filters that stores fingerprints in a cuckoo hash table.
 * Supports deletion while using less space than counting Bloom filters, and provides
 * better lookup performance in many scenarios.
 *
 * <p><strong>Algorithm:</strong>
 * <ul>
 *   <li>Stores fingerprints (partial hashes) in buckets</li>
 *   <li>Each item can reside in one of two candidate buckets</li>
 *   <li>Insert: place fingerprint in either bucket, or relocate existing items</li>
 *   <li>Delete: remove fingerprint from its bucket</li>
 *   <li>Query: check if fingerprint exists in either candidate bucket</li>
 * </ul>
 *
 * <p><strong>Use cases:</strong>
 * <ul>
 *   <li>Network deduplication with removal support</li>
 *   <li>Database query optimization</li>
 *   <li>Distributed caching with eviction</li>
 *   <li>Real-time spam filtering</li>
 * </ul>
 *
 * <p><strong>Advantages over Bloom Filters:</strong>
 * <ul>
 *   <li>Supports deletion (standard Bloom does not)</li>
 *   <li>More space-efficient at low FPR (&lt; 3%)</li>
 *   <li>Faster lookups (better cache locality)</li>
 *   <li>Can report when filter is full</li>
 * </ul>
 *
 * <p><strong>Limitations:</strong>
 * <ul>
 *   <li>Has a maximum capacity (insertion can fail)</li>
 *   <li>Duplicate insertions can cause issues</li>
 *   <li>FPR depends on fingerprint size</li>
 * </ul>
 *
 * <p><strong>Example:</strong>
 * <pre>
 * try (CuckooFilter cf = new CuckooFilter(100000)) {
 *     // Insert items
 *     cf.insert("user@example.com".getBytes());
 *
 *     // Check membership
 *     if (cf.contains("user@example.com".getBytes())) {
 *         System.out.println("Email might be registered");
 *     }
 *
 *     // Remove item (e.g., user unsubscribed)
 *     cf.remove("user@example.com".getBytes());
 * }
 * </pre>
 *
 * @see BloomFilter
 * @see CountingBloomFilter
 * @see <a href="https://www.cs.cmu.edu/~dga/papers/cuckoo-conext2014.pdf">Cuckoo Filter Paper</a>
 */
public final class CuckooFilter extends NativeSketch implements MergeableSketch<CuckooFilter> {

    private final long capacity;

    /**
     * Create a new CuckooFilter.
     *
     * @param capacity maximum number of items the filter can hold
     *                 The actual capacity may be slightly higher due to bucket sizing
     * @throws IllegalArgumentException if capacity &lt;= 0
     * @throws RuntimeException if native memory allocation fails
     */
    public CuckooFilter(long capacity) {
        if (capacity <= 0) {
            throw new IllegalArgumentException("capacity must be positive, got: " + capacity);
        }

        this.capacity = capacity;
        this.nativePtr = SketchOxideNative.cuckoofilter_new(capacity);

        if (this.nativePtr == 0) {
            throw new RuntimeException("Failed to allocate CuckooFilter");
        }
    }

    /**
     * Private constructor for deserialization.
     */
    private CuckooFilter(long nativePtr, long capacity) {
        this.nativePtr = nativePtr;
        this.capacity = capacity;
    }

    /**
     * Insert an item into the filter.
     *
     * Computes a fingerprint and places it in one of two candidate buckets.
     * If both buckets are full, existing fingerprints may be relocated
     * (cuckoo hashing).
     *
     * <p><strong>Note:</strong> Insertion can fail if the filter is too full.
     * The filter operates best below 95% capacity.
     *
     * @param item the item to insert (as raw bytes)
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public void insert(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        SketchOxideNative.cuckoofilter_insert(nativePtr, item);
    }

    /**
     * Insert a string item into the filter.
     *
     * @param item the string to insert
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public void insert(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        insert(item.getBytes());
    }

    /**
     * Remove an item from the filter.
     *
     * Removes the fingerprint from its bucket. Unlike CountingBloomFilter,
     * this is a true removal - the item will definitely not be found
     * after removal (unless it was inserted multiple times).
     *
     * <p><strong>Warning:</strong> Removing items that were never inserted
     * can cause false negatives. Only remove items you are certain were inserted.
     *
     * @param item the item to remove (as raw bytes)
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public void remove(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        SketchOxideNative.cuckoofilter_remove(nativePtr, item);
    }

    /**
     * Remove a string item from the filter.
     *
     * @param item the string to remove
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public void remove(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        remove(item.getBytes());
    }

    /**
     * Check if an item might be in the filter.
     *
     * Returns true if a matching fingerprint is found in either candidate bucket.
     * Returns false if no matching fingerprint exists (definitely not in set).
     *
     * @param item the item to check (as raw bytes)
     * @return true if the item might be in the set, false if definitely not
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public boolean contains(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return SketchOxideNative.cuckoofilter_contains(nativePtr, item);
    }

    /**
     * Check if a string item might be in the filter.
     *
     * @param item the string to check
     * @return true if the item might be in the set, false if definitely not
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public boolean contains(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return contains(item.getBytes());
    }

    /**
     * Merge another CuckooFilter into this one.
     *
     * <p><strong>Note:</strong> CuckooFilter merge is not typically supported
     * due to the nature of cuckoo hashing. This operation may fail or have
     * undefined behavior.
     *
     * @param other the filter to merge into this one
     * @throws UnsupportedOperationException always, as merge is not supported
     */
    @Override
    public void merge(CuckooFilter other) {
        throw new UnsupportedOperationException(
                "CuckooFilter does not support merge operations");
    }

    /**
     * Serialize the filter to binary format.
     *
     * @return binary representation of the filter
     * @throws IllegalStateException if the filter has been closed
     */
    public byte[] serialize() {
        checkAlive();
        return SketchOxideNative.cuckoofilter_serialize(nativePtr);
    }

    /**
     * Deserialize a CuckooFilter from binary format.
     *
     * @param data the serialized data from {@link #serialize()}
     * @return a new CuckooFilter instance
     * @throws IllegalArgumentException if data is invalid or corrupt
     * @throws NullPointerException if data is null
     */
    public static CuckooFilter deserialize(byte[] data) {
        if (data == null) {
            throw new NullPointerException("data cannot be null");
        }

        long ptr = SketchOxideNative.cuckoofilter_deserialize(data);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid or corrupt serialized CuckooFilter data");
        }

        return new CuckooFilter(ptr, 100);
    }

    /**
     * Get the capacity of the filter.
     *
     * @return the maximum number of items this filter can hold
     */
    public long getCapacity() {
        return capacity;
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            freeNative();
            nativePtr = 0;
        }
    }

    @Override
    protected void freeNative() {
        if (nativePtr != 0) {
            SketchOxideNative.cuckoofilter_free(nativePtr);
        }
    }

    @Override
    public String toString() {
        if (nativePtr != 0) {
            return String.format("CuckooFilter(capacity=%d)", capacity);
        }
        return "CuckooFilter(closed)";
    }
}

File: dotnet/SketchOxide/src/Membership/CuckooFilter.cs
using System;
using System.Text;
using SketchOxide.Native;

namespace SketchOxide.Membership;

/// <summary>
/// Cuckoo Filter for probabilistic set membership testing with deletion support.
///
/// A space-efficient probabilistic data structure that supports adding, removing,
/// and testing elements for set membership. Uses cuckoo hashing with fingerprints
/// for better space efficiency than counting Bloom filters.
/// </summary>
/// <remarks>
/// <para>
/// The Cuckoo filter provides:
/// - Support for element deletion (unlike standard Bloom filter)
/// - Better space efficiency than Counting Bloom filter
/// - Insert may fail if filter is too full (returns false)
/// - Lookup and delete operations are always successful
/// - Approximately 95% load factor before insertions fail
/// </para>
/// <para>
/// Ideal for applications where:
/// - Element deletion is required
/// - Space efficiency is important
/// - Insert failures can be handled gracefully
/// </para>
/// </remarks>
public sealed class CuckooFilter : NativeSketch
{
    private readonly ulong _size;

    /// <summary>
    /// Creates a new Cuckoo filter with the specified capacity.
    /// </summary>
    /// <param name="size">Maximum number of elements the filter can hold. Must be greater than 0.</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if size is 0.</exception>
    /// <exception cref="OutOfMemoryException">Thrown if native memory allocation fails.</exception>
    public CuckooFilter(ulong size)
    {
        if (size == 0)
            throw new ArgumentOutOfRangeException(nameof(size), size, "Size must be greater than 0");

        _size = size;
        NativePtr = SketchOxideNative.cuckoofilter_new(size);

        if (NativePtr == 0)
            throw new OutOfMemoryException("Failed to allocate native CuckooFilter");
    }

    /// <summary>
    /// Private constructor for deserialization.
    /// </summary>
    private CuckooFilter(ulong size, nuint ptr)
    {
        _size = size;
        NativePtr = ptr;
    }

    /// <summary>
    /// Gets the capacity of this filter.
    /// </summary>
    public ulong Capacity
    {
        get
        {
            CheckAlive();
            return _size;
        }
    }

    /// <summary>
    /// Inserts an element into the filter.
    /// </summary>
    /// <param name="data">The bytes representing the element to insert.</param>
    /// <returns>
    /// <c>true</c> if the element was successfully inserted;
    /// <c>false</c> if the filter is too full and insertion failed.
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public bool Insert(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                return SketchOxideNative.cuckoofilter_insert(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Inserts a string element into the filter.
    /// </summary>
    /// <param name="value">The string to insert.</param>
    /// <returns>
    /// <c>true</c> if the element was successfully inserted;
    /// <c>false</c> if the filter is too full and insertion failed.
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public bool Insert(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        return Insert(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Tests whether an element might be in the set.
    /// </summary>
    /// <param name="data">The bytes representing the element to query.</param>
    /// <returns>
    /// <c>true</c> if the element might be in the set (may be a false positive);
    /// <c>false</c> if the element is definitely not in the set (no false negatives).
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public bool Contains(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                return SketchOxideNative.cuckoofilter_contains(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Tests whether a string element might be in the set.
    /// </summary>
    /// <param name="value">The string to query.</param>
    /// <returns>
    /// <c>true</c> if the element might be in the set (may be a false positive);
    /// <c>false</c> if the element is definitely not in the set (no false negatives).
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public bool Contains(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        return Contains(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Removes an element from the filter.
    /// </summary>
    /// <param name="data">The bytes representing the element to remove.</param>
    /// <returns>
    /// <c>true</c> if the element was found and removed;
    /// <c>false</c> if the element was not found in the filter.
    /// </returns>
    /// <remarks>
    /// Unlike Counting Bloom filters, Cuckoo filters can safely handle remove
    /// operations for elements that may not have been inserted, though this
    /// could still cause issues if the fingerprint matches another element.
    /// </remarks>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public bool Remove(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                return SketchOxideNative.cuckoofilter_remove(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Removes a string element from the filter.
    /// </summary>
    /// <param name="value">The string to remove.</param>
    /// <returns>
    /// <c>true</c> if the element was found and removed;
    /// <c>false</c> if the element was not found in the filter.
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public bool Remove(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        return Remove(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Serializes the filter to a byte array.
    /// </summary>
    /// <returns>Serialized filter bytes.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public byte[] Serialize()
    {
        CheckAlive();
        return SketchOxideNative.cuckoofilter_serialize(NativePtr, out _);
    }

    /// <summary>
    /// Deserializes a Cuckoo filter from a byte array.
    /// </summary>
    /// <param name="data">Serialized filter bytes.</param>
    /// <param name="size">The size parameter used when creating the original filter.</param>
    /// <returns>A new CuckooFilter instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ArgumentException">Thrown if data is invalid.</exception>
    public static CuckooFilter Deserialize(byte[] data, ulong size)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        nuint ptr = SketchOxideNative.cuckoofilter_deserialize(data, (ulong)data.Length);
        if (ptr == 0)
            throw new ArgumentException("Failed to deserialize CuckooFilter: invalid data");

        return new CuckooFilter(size, ptr);
    }

    /// <summary>
    /// Returns a string representation of the filter.
    /// </summary>
    public override string ToString()
    {
        if (IsDisposed)
            return "CuckooFilter(disposed)";
        return $"CuckooFilter(capacity={_size})";
    }

    /// <summary>
    /// Frees the native Rust instance.
    /// </summary>
    protected override void FreeNative()
    {
        if (NativePtr != 0)
        {
            SketchOxideNative.cuckoofilter_free(NativePtr);
            NativePtr = 0;
        }
    }
}
